### PR TITLE
[Ts] Migrate batch_exchange.js

### PR DESCRIPTION
**Pull request by fleupold**
To make this seemingly huge change reviewable, it is split into two commits (plus additional ones for addressing comments likely to come). The first commit just moves `test/batch_exchange.js` to `test/BatchExchange.spec.ts` without changing any line of code.

The second then refactors `test/BatchExchange.spec.ts` to be valid typescript (which due to the way formatting works touches almost all of its 3000 lines). By just looking at the second commit, we can at least see the file and changes side by side.

I didn't change any testing logic. Most changes are around comparing BN with BN instead of with numbers or strings (added benefit of strict type-checking). The other change is that most unnecessary `.call()` invocations are now omitted. E.g.

```ts
BATCH_TIME = (await batchExchange.BATCH_TIME.call()).toNumber()
=>
BATCH_TIME = (await batchExchange.BATCH_TIME()).toNumber()
```

I also added a `toOrderResult` helper function that takes the converts the assumed tuple returned type (`[BN, BN, BN, BN, BN]`) from `orders` and turns it into a named typed. This can be removed once https://github.com/ethereum-ts/TypeChain/issues/232 is fixed.

Lastly, it adds type-annotations for linking libraries into the contract (the current typings assume this to be only done through the `Deployer` object that is injected to the migrations script and moves the remaining utility functions into `test/utilities.ts`.

### Test Plan

CI

### CODE REVIEW

**Review by e00E — APPROVED** [review]
I have only read the code that isn't in `test/BatchExchange.spec.ts `.
For the test we can see that coverage remains at 100% and I checked that all of the batch exchange tests are still there using the travis logs. Imo that's enough and I'd like to avoid reading that whole file.

File: test/utilities.ts [hunk 1]
```diff
@@ -50,14 +52,17 @@ export async function closeAuction(
...
 }
 
-export async function sendTxAndGetReturnValue(
+export async function sendTxAndGetReturnValue<T>(
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  method: any,
+  method: {
+    sendTransaction: (...args: any[]) => Promise<string>;
+    call: (...args: any[]) => Promise<T>;
+  },
```
Comment (nlordell) [inline test/utilities.ts:60]: Awesome! More type safety! :tada:

**Review by nlordell — APPROVED** [review]
Looks great! I only read about 10% of the batch_exchange test file and it looks like mostly reformatting, adding semicolons and adding types.

As @e00E the coverage numbers and actually ran tests seem to indicate everything is still working.

File: test/stablex_utils.js [hunk 2]
```diff
@@ -1,78 +0,0 @@
...
-  const instance = await BatchExchange.new(maxTokens, feeToken.address)
-  const tokens = [feeToken]
-  for (let i = 0; i < numTokens - 1; i++) {
-    const token = await MockContract.new()
-    await instance.addToken(token.address)
-    await token.givenAnyReturnBool(true)
-    tokens.push(token)
-  }
-  return instance
-}
```
Comment (bh2smith) [inline test/stablex_utils.js:33]: Note: It appears this had already been migrated and was just deleted here.

**Review by bh2smith — APPROVED** [review]
Hot diggity! Great stuff!

File: test/BatchExchange.spec.ts [hunk 3]
```diff
@@ -0,0 +1,3000 @@
...
+        solution.prices,
+        solution.tokenIdsForPrice,
+        { from: solver },
+      );
+
+      assert(
+        objectiveValue.gtn(0),
+        "the computed objective value is greater than 0",
+      );
+      assert(objectiveValue.eq(solution.objectiveValue));
```
Comment (bh2smith) [inline test/BatchExchange.spec.ts:638]: Including both assertions for context:

The first assertion here does not seem necessary as it is trumped by the second. If we do chose to keep it, however, I think the failure string should be modified to its negation?

**Pull request merged**